File: backend/app/api/v1/map.py
"""Map data endpoints.

Kept separate from the complaint list because a map has different needs: a
viewport bounding box instead of pages, a trimmed payload, and a hard cap on
how many markers can come back at once.
"""

from __future__ import annotations

from datetime import datetime

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps import get_optional_user, get_session
from app.api.serializers import to_map_issue
from app.core.enums import ComplaintStatus, DamageType, PriorityLevel, UserRole
from app.models.user import User
from app.repositories.complaint import ComplaintFilters, ComplaintRepository
from app.schemas.complaint import MapIssue

router = APIRouter(prefix="/map", tags=["map"])

#: Hidden from anonymous and citizen viewers of the public map.
_PUBLIC_HIDDEN = {ComplaintStatus.REJECTED, ComplaintStatus.DUPLICATE}


@router.get("/issues", response_model=list[MapIssue], summary="Map markers")
def map_issues(
    min_lat: float | None = Query(None, ge=-90, le=90),
    min_lon: float | None = Query(None, ge=-180, le=180),
    max_lat: float | None = Query(None, ge=-90, le=90),
    max_lon: float | None = Query(None, ge=-180, le=180),
    damage_type: list[DamageType] | None = Query(None),
    priority_level: list[PriorityLevel] | None = Query(None),
    status_filter: list[ComplaintStatus] | None = Query(None, alias="status"),
    created_from: datetime | None = Query(None),
    created_to: datetime | None = Query(None),
    exclude_closed: bool = Query(False),
    limit: int = Query(1000, ge=1, le=5000),
    user: User | None = Depends(get_optional_user),
    db: Session = Depends(get_session),
) -> list[MapIssue]:
    is_staff = user is not None and user.role in (UserRole.ADMIN, UserRole.REPAIR_TEAM)

    statuses = status_filter
    if not is_staff:
        allowed = [s for s in (status_filter or list(ComplaintStatus)) if s not in _PUBLIC_HIDDEN]
        statuses = allowed

    bbox = None
    if None not in (min_lat, min_lon, max_lat, max_lon):
        # Normalise, so a viewport dragged in either direction still works.
        bbox = (
            min(min_lat, max_lat),
            min(min_lon, max_lon),
            max(min_lat, max_lat),
            max(min_lon, max_lon),
        )

    filters = ComplaintFilters(
        status=statuses,
        damage_type=damage_type,
        priority_level=priority_level,
        created_from=created_from,
        created_to=created_to,
        exclude_closed=exclude_closed,
        bbox=bbox,
    )
    return [
        to_map_issue(item, precise=is_staff)
        for item in ComplaintRepository(db).map_points(filters, limit=limit)
    ]

File: supabase/functions/api/routes/map.ts
// Ported from backend/app/api/v1/map.py. Same public-visibility rule as the
// complaints feed (REJECTED/DUPLICATE hidden from anonymous and citizen
// viewers), but a lean projection with a viewport bounding box instead of
// pages, and a hard cap on marker count.

import { Hono } from "hono";
import { optionalUser } from "../_shared/auth.ts";
import { userClient } from "../_shared/supabase.ts";
import { toMapIssue } from "../_shared/serializers.ts";
import { mapPoints, type ComplaintFilters } from "../repositories/complaint.ts";
import type { ComplaintStatus, DamageType, PriorityLevel } from "../_shared/enums.ts";

export const map = new Hono();

const PUBLIC_HIDDEN = new Set<ComplaintStatus>(["REJECTED", "DUPLICATE"]);
const ALL_STATUSES: ComplaintStatus[] = [
  "PENDING", "AI_ANALYZED", "PRIORITIZED", "ASSIGNED", "IN_PROGRESS", "RESOLVED", "REJECTED", "DUPLICATE",
];

map.get("/issues", async (c) => {
  const req = c.req.raw;
  const url = new URL(req.url);
  const user = await optionalUser(req);
  const isStaff = user !== null && (user.role === "ADMIN" || user.role === "REPAIR_TEAM");

  const statusParam = url.searchParams.getAll("status") as ComplaintStatus[];
  const statuses = isStaff
    ? statusParam.length > 0 ? statusParam : undefined
    : (statusParam.length > 0 ? statusParam : ALL_STATUSES).filter((s) => !PUBLIC_HIDDEN.has(s));

  const minLat = url.searchParams.get("min_lat");
  const minLon = url.searchParams.get("min_lon");
  const maxLat = url.searchParams.get("max_lat");
  const maxLon = url.searchParams.get("max_lon");
  let bbox: [number, number, number, number] | undefined;
  if (minLat !== null && minLon !== null && maxLat !== null && maxLon !== null) {
    const a = Number(minLat), b = Number(minLon), x = Number(maxLat), y = Number(maxLon);
    bbox = [Math.min(a, x), Math.min(b, y), Math.max(a, x), Math.max(b, y)];
  }

  const createdFromParam = url.searchParams.get("created_from");
  const createdToParam = url.searchParams.get("created_to");
  const limit = Math.min(5000, Math.max(1, Number(url.searchParams.get("limit") ?? "1000")));

  const filters: ComplaintFilters = {
    status: statuses,
    damageType: url.searchParams.getAll("damage_type") as DamageType[],
    priorityLevel: url.searchParams.getAll("priority_level") as PriorityLevel[],
    createdFrom: createdFromParam ? new Date(createdFromParam) : undefined,
    createdTo: createdToParam ? new Date(createdToParam) : undefined,
    excludeClosed: url.searchParams.get("exclude_closed") === "true",
    bbox,
  };
  if (filters.damageType?.length === 0) filters.damageType = undefined;
  if (filters.priorityLevel?.length === 0) filters.priorityLevel = undefined;

  const items = await mapPoints(userClient(req), filters, limit);
  return c.json(items.map(toMapIssue));
});
